cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,382 @@
+        }
+
+    def _parameter_names_(self) -> AbstractSet[str]:
+        return {
+            name
+            for symbol in protocols.parameter_symbols(self.circuit)
+            for name in protocols.parameter_names(
+                protocols.resolve_parameters(symbol, self.param_resolver)
+            )
+        }
+
+    def _decompose_(self) -> 'cirq.OP_TREE':
+        result = self.circuit.unfreeze()
+        result = result.transform_qubits(lambda q: self.qubit_map[q])
+        if self.repetitions < 0:
+            result = result ** -1
+        result = protocols.with_measurement_key_mapping(result, self.measurement_key_map)
+        result = protocols.resolve_parameters(result, self.param_resolver)
+
+        return list(result.all_operations()) * abs(self.repetitions)
+
+    # Methods for string representation of the operation.
+
+    def __repr__(self):
+        result = f'cirq.CircuitOperation({self.circuit!r})'
+        base_op = self.base_operation()
+
+        def dict_repr(d: Dict) -> str:
+            pairs = [f'    {proper_repr(k)}: {proper_repr(v)},' for k, v in sorted(d.items())]
+            return '\n'.join(['{', *pairs, '}'])
+
+        if self.qubit_map != base_op.qubit_map:
+            result += f'.with_qubit_mapping({dict_repr(self.qubit_map)})'
+        if self.measurement_key_map != base_op.measurement_key_map:
+            result += f'.with_measurement_key_mapping({dict_repr(self.measurement_key_map)})'
+        if self.param_resolver != base_op.param_resolver:
+            result += f'.with_params({proper_repr(self.param_resolver)})'
+        if self.repetitions != 1:
+            result += f'.repeat({self.repetitions})'
+        return result
+
+    def __str__(self):
+        # TODO: support out-of-line subcircuit definition in string format.
+        header = self.circuit.serialization_key() + ':'
+        msg_lines = str(self.circuit).split('\n')
+        msg_width = max([len(header) - 4] + [len(line) for line in msg_lines])
+        circuit_msg = '\n'.join(
+            ['[ {line:<{width}} ]'.format(line=line, width=msg_width) for line in msg_lines]
+        )
+        args = []
+        base_op = self.base_operation()
+
+        def dict_str(d: Dict) -> str:
+            pairs = [f'{k}: {v}' for k, v in sorted(d.items())]
+            return '{' + ', '.join(pairs) + '}'
+
+        if self.qubit_map != base_op.qubit_map:
+            args.append(f'qubit_map={dict_str(self.qubit_map)}')
+        if self.measurement_key_map != base_op.measurement_key_map:
+            args.append(f'key_map={dict_str(self.measurement_key_map)}')
+        if self.param_resolver != base_op.param_resolver:
+            args.append(f'params={self.param_resolver.param_dict}')
+        if self.repetitions != 1:
+            args.append(f'loops={self.repetitions}')
+        if not args:
+            return f'{header}\n{circuit_msg}'
+        return f'{header}\n{circuit_msg}({", ".join(args)})'
```

Comment:
Another diagram nit:
It's great to have the qubit map there. However, we could make it a bit cleaner.

Code:

```
import cirq
from cirq import *

a, b, c, d = GridQubit.rect(2,2)
q = [NamedQubit(i) for i in ["a","b","c"]]
op = CircuitOperation(FrozenCircuit(cirq.X(q[0]), cirq.Y(q[1]), cirq.Z(q[2]), cirq.CX(q[0], q[1]))).with_tags(["tag"])
print(Circuit(X(a), Y(b), Z(c), CX(a, b), H(c), op.with_qubits(a,b,d), [H(i) for i in [a,b,c,d]]))
```

Current output:

```
                   ┌──────────────────────────────────────────────────────────────────────────────────┐
                    Circuit_0x740eb7df41d7eca6:
                    [ a: ───X───@───          ]
(0, 0): ───X───@────[           │             ]───────────────────────────────────────────────────────────H───
               │    [ b: ───Y───X───          ]
               │    [                         ]
               │    [ c: ───Z───────          ](qubit_map={a: (0, 0), b: (0, 1), c: (1, 1)})[['tag']]
               │    │
(0, 1): ───Y───X────#2────────────────────────────────────────────────────────────────────────────────────H───
                    │
(1, 0): ───Z───H────┼────────────────────────────────────────────────────────────────────────────────H────────
                    │
(1, 1): ────────────#3────────────────────────────────────────────────────────────────────────────────────H───
                   └──────────────────────────────────────────────────────────────────────────────────┘
```

Improved: 
1. mapping in-place for qubits outer -> inner
2. tag(s) to the header instead 

```
                   ┌─────────────────────────────────────┐        
                    Circuit_0x740eb7df41d7eca6[['tag']]:          
                    [ (0,0) -> a: ───X───@───]                    
(0, 0): ───X───@────[                    │   ]──────────────H───  
               │    [ (0,1) -> b: ───Y───X───]                    
               │    [                        ]                    
               │    [ (1,1) -> c: ───Z───────]                    
               │    │                                             
(0, 1): ───Y───X────#2───────────────────────────────────────H─── 
                    │                                             
(1, 0): ───Z───H────┼───────────────────────────────────H──────── 
                    │                                             
(1, 1): ────────────#3───────────────────────────────────────H─── 
                   └─────────────────────────────────────┘        
```

Reply:
If you'd like to see this reflected in the final subcircuit diagram format, please copy this comment to #3605.